torch/csrc/autograd/engine.cpp

```diff
@@ -742,44 +752,134 @@ auto Engine::execute(const edge_list& roots,
   return execute_with_graph_task(graph_task, graph_root)->wait();
 }
 
-void Engine::initialize_threads_pool() {
+void Engine::initialize_device_threads_pool() {
   track_bad_autograd_forks();
   TORCH_CHECK(!in_bad_autograd_fork,
               "Unable to handle autograd's threading in combination with fork-based multiprocessing. "
               "See https://github.com/pytorch/pytorch/wiki/Autograd-and-Fork");
-  std::call_once(start_threads_flag_, &Engine::start_threads, this);
+  std::call_once(start_device_threads_flag_, &Engine::start_device_threads, this);
 }
 
 void Engine::enqueue_blocked_task_on_cpu(NodeTask task) {
-  initialize_threads_pool();
-  ready_queue(at::kCPU).push(
+  initialize_device_threads_pool();
+  std::shared_ptr<GraphTask> graph_task = task.base_.lock();
+  // The graph_task must be alive at this point, because internal autograd machinary
+  // who calls this API (Distributed Autograd Engine) increases outstanding_tasks_
+  // outside this API to keep the GraphTask alive
+  TORCH_INTERNAL_ASSERT(graph_task, "GraphTask is no longer valid!");
+  ready_queue(graph_task, at::kCPU).push(
       std::move(task), /* incrementOutstandingTasks */ false);
+  std::unique_lock<std::mutex> lock(print_mutex_);
+  LOG(INFO)<<"enqueued blocked task on CPU ready queue for graphTask: "<< graph_task;
+}
+
+std::shared_ptr<GraphTask> Engine::execute_node_task(const std::unique_ptr<NodeTask>& task) {
+  // local_graph_task represents the graph_task we retrieve from the queue.
+  // The outer graph_task represents the overall graph_task we need to execute
+  // for reentrant execution.
+  std::shared_ptr<GraphTask> local_graph_task;
+  if (!(local_graph_task = task->base_.lock())) {
+    // Reentrant thread's graph task should not expire since we hold a
+    // reference to it in this method.
+    LOG(INFO) << "GraphTask for function " << task->fn_->name()
+              << " is no longer valid, skipping execution";
+    return local_graph_task;
+  }
+
+  if (task->fn_ && !local_graph_task->has_error_.load()) {
+    AutoGradMode grad_mode(local_graph_task->grad_mode_);
+    try {
+      evaluate_function(local_graph_task, task->fn_.get(), task->inputs_);
+    } catch (std::exception& e) {
+      thread_on_exception(local_graph_task, task->fn_, e);
+    }
+  }
+
+  // Decrement the outstanding tasks.
+  --local_graph_task->outstanding_tasks_;
+
+  return local_graph_task;
+}
+
+void Engine::execute_until_ready_queue_empty(const std::shared_ptr<GraphTask>& graph_task) {
```

Comment:
This function name is confusing to me.
It will lock until the ready queue is empty + use the aten thread pool to finish the executation later.

It actually make sure the whole graph_task is executed. Not just until the ready_queue is empty.